### ISSUE TITLE
Native vim support for displaying search match index like `(n of N)`

### ISSUE THREAD

**Issue by dbarnett** [vim/vim#453]
When iterating through search matches with the `n` mapping, could vim display the match count and index of the current match in the list of matches? Instead of just displaying `/in` in the message area when you search for `in`, it could display `[1/7]  /in`, `[2/7]  /in`, etc. as you step through.

There are plugins https://github.com/google/vim-searchindex and https://github.com/henrik/vim-indexed-search that implement this somewhat effectively in vimscript, but they have a lot of quirks to work around.

**Comment by advocateddrummer** [comment]
+1

This would be pretty great!

**Comment by mattn** [comment]
for example, after type `/test`, you will expect `[1/3]` maybe.

```
test1
test2
test3
```

`nn` will be work fine. then you append `test4` into the buffer. what you expect?

**Comment by dbarnett** [comment]
Messages in the message area disappear when you enter insert mode or do any other buffer operations. They only reappear the next time you trigger `n` or other search operations, at which point we'd have the updated count (this already works as expected in https://github.com/google/vim-searchindex).

**Comment by mattn** [comment]
When the count of indexes are little, It may works pretty good. However vim must update the count after exiting insert mode every time. So if the count will be huge, some users may stop to use vim.

**Comment by rburny** [comment]
@mattn I agree that this feature has potential performance implications and care would be needed to avoid visible lags. But like many advanced Vim features, such feature should be highly configurable (by default it could be turned off, or limited to some file size), so that users would be able to disable it.

To give perspective, here are some measurements for https://github.com/google/vim-searchindex on a 1.1 MB, 30 kLOC file. Perhaps it could be faster if integrated natively into Vim:
simple text search: 0.03 sec
regex search: 0.1 sec   (obviously, this depends on the pattern)

As you can see, if you keep your files to reasonable size - which is a good practice in coding anyway! - search counting can work pretty fast. But for big files, an option to disable it would be necessary.

**Comment by vim-ml** [comment]
On Mo, 19 Okt 2015, Radoslaw Burny wrote:

> @mattn I agree that this feature has potential performance implications and care would be needed to avoid visible lags. But like many advanced Vim features, such feature should be highly configurable (by default it could be turned off, or limited to some file size), so that users would be able to disable it.
> 
> To give perspective, here are some measurements for https://github.com/google/vim-searchindex on a 1.1 MB, 30 kLOC file. Perhaps it could be faster if integrated natively into Vim:
> simple text search: 0.03 sec
> regex search: 0.1 sec   (obviously, this depends on the pattern)

What file and what pattern did you test this on?
I expect a reasonable performance penalty simply for keep on searching 
to count the number of patterns. So if this gets implemented, it should 
be definitely be allowed to disabled. Perhaps using the  shortmess 
option?

 (I might have a preliminary patch for that, but I like to get some 
 performance testing).

Best,

## Christian

F: Warum trinken Elefanten keine Martinis?
A: Schon mal versucht, eine Olive aus der Nase zu bekommen?

**Comment by dbarnett** [comment]
+1 to having it behind an option, if only because the behavior change might surprise people. My vote would be to have it opt-in for a few months and then consider enabling it by default.

If you're really concerned about performance, you could try some tricks with caching, displaying `[1/?]` for the first search, finishing expensive searches when vim is idle, etc.

**Comment by chrisbra** [comment]
Okay, here is something to play with: https://github.com/chrisbra/vim-mq-patches/blob/master/search_stat

**Comment by mattn** [comment]
https://github.com/chrisbra/vim-mq-patches/blob/master/search_stat#L178
please check ui_breakcheck and got_int.

https://github.com/chrisbra/vim-mq-patches/blob/master/search_stat#L197
check NULL first, or move STRCPY into if-block.

**Comment by chrisbra** [comment]
@mattn Thanks for the feedback.
I have updated the patch somemore (caches value, should work correclty with 'rl' set, when truncating the output and when wrapping around). However, whenever the cached value needs to be updated, the delay is noticeable. Especially with a long file and a pattern that matches often and although I like that Vim displays the statistics, I find the delay not acceptable.

**Comment by dbarnett** [comment]
Interesting. I'd still like to have it as an option, and personally I'd pbb enable it by default and use an autocmd to selectively disable it for long files (global+local setting would be ideal).

**Comment by rburny** [comment]
To reply to the previous question about my measurements: I took one of the bigger files from my team's C++ codebase, and copied it 5 times. The result was 1.1 MB, 30 kLOC C++ file. Only the first search time was measured; vim-searchindex caches the results, so repeated search would be immediate. I did two types of experiments:
- searches for short and common words (so that many matches are hit), e.g. "void" or identifiers widely used in the file. Number of matches was up to 10k, depending on the pattern, and didn't affect the timing much. The counting took 30ms.
- simple regex like `a.*b.*(` . This is something I would use on a one-off basis if I don't remember exact method name; obviously, this only exposes tiny part of complexity of regexes. In this case, number of matches was 2k, and the result was 100ms.

@chrisbra Christian, what results did you achieve? How does it scale with the number of lines & number of matches? If we could achieve a reasonably small timeout for files below 10kLOC and 1000 matches, this would already be a huge win (as I think it covers "normal" programming usage).

One approach to limiting lag  would be to have time-based "deadline" and just try to count the matches on the best-effort basis.

**Comment by vim-ml** [comment]
2015-10-22 2:29 GMT+03:00 Radoslaw Burny vim-dev-github@256bit.org:

> To reply to the previous question about my measurements: I took one of the
> bigger files from my team's C++ codebase, and copied it 5 times. The result
> was 1.1 MB, 30 kLOC C++ file. Only the first search time was measured;
> vim-searchindex caches the results, so repeated search would be immediate.
> I did two types of experiments:
> - searches for short and common words (so that many matches are hit),
>   e.g. "void" or identifiers widely used in the file. Number of matches was
>   up to 10k, depending on the pattern, and didn't affect the timing much. The
>   counting took 30ms.
> - simple regex like a._b._( . This is something I would use on a
>   one-off basis if I don't remember exact method name; obviously, this only
>   exposes tiny part of complexity of regexes. In this case, number of matches
>   was 2k, and the result was 100ms.
> 
> Christian, what results did you achieve? How does it scale with the number
> of lines & number of matches? If we could achieve a reasonably small
> timeout for files below 10kLOC and 1000 matches, this would already be a
> huge win (as I think it covers "normal" programming usage).

​I think that most problematic will be multi-line matches (e.g. something
like `^\S.*(\_[^)]*​\<foo,` to find functions which have “foo” argument
(example was not actually tested and depends on the coding style)). Also
huge slowdown will be present in case of using back and forward references
(especially with a new regexp engine), but this is less critical with the
existence of \zs and \ze which cover many use-cases.

Most my searches though are simple `\<word\>` or `word`, which are fast.

**Comment by lilydjwg** [comment]
Nobody mentioned logs? I usually view logs with Vim. They are large, sometimes hundreds of megabytes. And in those huge files I search a lot. I hope this feature won't make it slow.

**Comment by chrisbra** [comment]
@lilydjwg If this feature gets integrated, there will be a switch to turn it on. Currently in my patch, it is disabled by default and you need to set shortmess option.

**Comment by brammool** [comment]
David Barnett wrote:

> +1 to having it behind an option, if only because the behavior change
> might surprise people. My vote would be to have it opt-in for a few
> months and then consider enabling it by default.
> 
> If you're really concerned about performance, you could try some
> tricks with caching, displaying `[1/?]` for the first search,
> finishing expensive searches when vim is idle, etc.

There could be an option that disables the feature when the file is
longer a certain number of lines and/or when some number of matches
are found.

Also, it should probably only display the count after performing the
search, not for incremental search.  And only do it when the command was
typed.

Considering that we have 'hlsearch', which I assume many people have
enabled, searching for more matches shouldn't slow down normal editing
much.

## 

hundred-and-one symptoms of being an internet addict:
60. As your car crashes through the guardrail on a mountain road, your first
    instinct is to search for the "back" button.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by bohrshaw** [comment]
> And only do it when the command was typed.

So there would be no count update on `n` or `N`, which IMO is sensible.
We could then update the count with `/<CR>`.

**Comment by chdiza** [comment]
@chrisbra posted a patch to the mailing list in which the match index is ON by default.

This is a plea to leave it off by default, which seemed to be the plan in this thread.

**Comment by lifecrisis** [comment]
One thing that needs to be well-defined is when the message will "go away".  Currently in [vim-searchindex](https://www.github.com/google/vim-searchindex), switching windows will leave the message at the bottom of the screen. Does the index message apply in the new window? Of course not... it should probably disappear as soon as the the cursor moves away from the match.  That is, it should only be visible while the user is actively cycling through matches.

**Comment by joeky888** [comment]
Ideas from [here](https://stackoverflow.com/a/23724875) and [here](https://stackoverflow.com/a/4671105)

Notice the numbers on the statusline [Nth of N]

![GIF](http://i.imgur.com/2RFvbJF.gif)

## code
```
function! SearchCount()
  let keyString=@/
  let pos=getpos('.')
  try
    redir => nth
      silent exe '0,.s/' . keyString . '//ne'
    redir => cnt
      silent exe '%s/' . keyString . '//ne'
    redir END
    return matchstr( nth, '\d\+' ) . '/' . matchstr( cnt, '\d\+' )
  finally
    call setpos('.', pos)
  endtry
endfunction
set statusline+=[%{SearchCount()}] " Nth of N when searching
```

Add this code to your vimrc file.

**Comment by chrisbra** [comment]
there are some issues with your vimscript snippet (not ending the previous redir, not using `g` flag for `:%s` command, not sure the first index is correct (how do you know whether the cursor is before or after the match at the current line?) Also this will probably slow down Vim a lot, as the statusline is evaluated a lot. At least some kind of caching should be considered, before throwing this into the statusline.

**Comment by joeky888** [comment]
How about use another function, say UpdateSearch(), do an update after searching？

## Code
```
function! SearchCount()
  " variable b:lastKey = [key, Nth, N]
  if !exists('b:lastKey') | let b:lastKey = [[], 0, 0] | endif
  return b:lastKey[1] . "/" . b:lastKey[2]
endfunction

function! UpdateSearch()
  let pos=getpos('.')
  let key=[@/, b:changedtick]

  let b:lastKey[0] = key

  redir => cnt
    silent exe '%s/' . key[0] . '//gne'
  redir END
  let b:lastKey[2] = str2nr(matchstr( cnt, '\d\+' ))

  redir => nth
    silent exe '0,.s/' . key[0] . '//gne'
  redir END

  let b:lastKey[1] = str2nr(matchstr( nth, '\d\+' ))
  call setpos('.', pos)
endfunction

function CommandAfterSearch()
  if getcmdtype() == '/'
    return "\<cr>:call UpdateSearch()\<cr>"
  else
    return "\<cr>"
  endif
endfunction

function NextSearch()
  let l:line = line(".")
  normal! n
  if l:line != line(".")
    let b:lastKey[1]=b:lastKey[1]+1 <= b:lastKey[2] ? b:lastKey[1]+1 : b:lastKey[1]+1-b:lastKey[2]
  endif
endfunction

function PreviousSearch()
  let l:line = line(".")
  normal! N
  if l:line != line(".")
    let b:lastKey[1]=b:lastKey[1]-1 > 0 ? b:lastKey[1]-1 : b:lastKey[1]-1+b:lastKey[2]
  endif
endfunction

nnoremap <silent> n :call NextSearch()<CR>
nnoremap <silent> N :call PreviousSearch()<CR>
cnoremap <silent> <expr> <Enter> CommandAfterSearch()
set statusline+=[%{SearchCount()}] " Nth of N when searching
```

**Comment by chrisbra** [comment]
for what its worth, i updated my [patch](http://github.com/chrisbra/vim-mq-patches/blob/master/search_stat) to Vim 8.0.0993 and made the feature opt-in as requested.

**Comment by dbarnett** [comment]
@chrisbra Looks like no further progress has been made getting that patch approved and submitted, right?

**Comment by chrisbra** [comment]
Well, FWIW, I am not sure what would be missing.

**Comment by liushapku** [comment]
I would like to vote for this patch. 

Despite the efficiency, the way the plugin `searchindex` does its job has side effects: the last replacement string and the last flags used for substitution are overwritten by the plugin. see this [issue](https://github.com/google/vim-searchindex/issues/18).

For example, in the following Ex command sequence (should to be typed, not running in a function)

    :s/red/green/g
    /blue
    :~

the last `:~` will replace `blue` with empty string. If the `:~` in the last snippet is changed to `:~&`, the flags used will be `eng` instead of `g`.

The problem for the last replacement string can be solved, while the problem for the losing flags can not be solved easily since `vim` has no way to keep the last flags unchanged; unless we add another feature `keepflags` as similar to `keeppattern`.

**Comment by Houl** [comment]
Maybe what's actually needed is a substitute command which avoids _any_ side effect.

**Comment by dbarnett** [comment]
@chrisbra Like everything's lined up and we're purely waiting on a decision from Bram about accepting the patch? Didn't look like it was merged into vim yet.

**Comment by liushapku** [comment]
> Maybe what's actually needed is a substitute command which avoids _any_ side effect.

Instead of a new command or function, I would rather suggest add a new flag (for example `k`) to `:s`.

It seems to be easier.

**Comment by joeky888** [comment]
Dying for this +1

I've used the snippet https://github.com/vim/vim/issues/453#issuecomment-324891452 for a while.

I don't use plugins since I often work on embed systems without git and net connection.

**Comment by chrisbra** [comment]
Updated the [patch for 8.1.576](https://github.com/chrisbra/vim-mq-patches/blob/master/search_stat). Major enhancement: Should work better with redraws now.
Also available at the search_stat branch here: https://github.com/chrisbra/vim/tree/search_stat

**Comment by tjkirch** [comment]
I'd love to see something like this in Vim.  I updated chrisbra's patch to latest Vim to try to help.  https://github.com/chrisbra/vim-mq-patches/pull/2

**Comment by chrisbra** [comment]
fixed by 9dfa3139198b38b28673e251a375643
